proof_test.go

```diff
@@ -80,8 +113,89 @@ func TestEthTrie(t *testing.T) {
 			if string(recovered) != tc.query {
 				t.Fatalf("Recovered key %s doesn't match query %s\n", string(recovered), tc.query)
 			}
+
+			// see if we can verify this proof
+			err = VerifyProof(proof, hash)
+			if err != nil {
+				t.Fatalf("Invalid proof %+v", err)
+			}
 		})
 
 	}
+}
+
+// TestRandomTrie is basically a fuzz-tester
+// If there is an error, it should dump out enough info to create a targetted case above
+func TestRandomTries(t *testing.T) {
+	runs := 20
+	size := 5000
+
+	for i := 0; i<runs; i++ {
+		t.Run(fmt.Sprintf("Run %d", i), func (t *testing.T) {
+			tr, keys := randomTrie(t, size)
+			// grab one of the last ones (random) to query
+			query := keys[len(keys)-3]
+
+			proof, err := ComputeProof(tr, query.k)
+			if err != nil {
+				t.Fatalf("ComputeProof: %+v", err)
+			}
+			t.Logf("Path length %d", len(proof.Steps))
+			if !bytes.Equal(query.v, proof.Value) {
+				t.Fatalf("invalid value: %X (expected %X)", proof.Value, query.v)
+			}
+
+			recovered := proof.RecoverKey()
+			if !bytes.Equal(query.k, recovered) {
+				t.Fatalf("Recovered key %X doesn't match query %X\n", recovered, query.k)
+			}
+
+			err = VerifyProof(proof, tr.Hash())
+			if err != nil {
+				t.Fatalf("Invalid proof %+v", err)
+			}
+		})
+	}
+}
+
+type kv struct {
+	k []byte
+	v []byte
+}
+
+func randomTrie(t *testing.T, n int) (*trie.Trie, []kv) {
+	db := ethdb.NewMemDatabase()
+	tr, err := trie.New(common.BytesToHash(nil), trie.NewDatabase(db))
+	if err != nil {
+		t.Fatalf("cannot create an empty trie: %s", err)
+	}
+
+	var vals []kv
+	for i := byte(0); i < 100; i++ {
```

Comment:
This is code from the ethereum codebase and I was not able to understand it. Do you know why the trie is initialized with the requested amount of values **plus** about 200 values more? Why is `value` overwriting at some point `value2` items (after 10 items)?

Reply:
No real idea why, I was just going to add random values, but figured to do what they do.

I finally got this stuff to run basically, now need to escape some days from ethereum to recover sanity